cpp/src/arrow/filesystem/s3fs_test.cc

```diff
@@ -1086,6 +1116,74 @@ TEST_F(TestS3FS, DeleteDirContentsAsync) {
   AssertFileInfo(infos[3], "bucket/somefile", FileType::File);
 }
 
+TEST_F(TestS3FS, DeleteFileWithMarker) {
+  FileSelector select;
+  select.base_dir = "bucket/somedir/subdir";
+  select.recursive = false;
+  std::vector<FileInfo> infos;
+
+  // Expected files before deletion
+  std::vector<std::string> expected_before = {
+      "somedir/subdir/subfile",
+  };
+
+  // Verify the list of objects before deletion
+  AssertListObjects(client_.get(), "bucket", "somedir/subdir", expected_before);
+
+  ASSERT_OK_AND_ASSIGN(infos, fs_->GetFileInfo(select));
+  ASSERT_EQ(infos.size(), 1);
+
+  ASSERT_OK(fs_->DeleteFile("bucket/somedir/subdir/subfile"));
+
+  // Expected files after deletion
+  std::vector<std::string> expected_after = {
+      "somedir/subdir/",
+  };
+
+  // Verify the list of objects after deletion
+  AssertListObjects(client_.get(), "bucket", "somedir/subdir", expected_after);
```

Comment:
Initially, I tried to write the test using only the `GetFileInfo` function, but I added the `ListObjectsV2Request` API because I thought it would more intuitively reproduce the issue.